### DISCUSSION TITLE
Completion bug in Astro files

### DISCUSSION THREAD

**Discussion by Trildar** [LazyVim/LazyVim#1455] [Q&A]
When inserting completions from Astro LSP in some situations, the completion seems to be inserted over the wrong range, usually leaving the first letter of what was typed to trigger completion suggestions and overwriting some of the existing text after. The case I notice this happening with is when inside an injection in the HTML template part of a .astro file, i.e. inside {} of something like `<div id={}>`, don't think I've encountered it happening in any other case. Also maybe worth noting is that selecting suggestions inserts correctly, but confirming the selection still corrupts the text afterwards.

I tried writing a minimal config to see if it was an issue with nvim-cmp and the Astro LSP, but it seems to be fine with that config. I have some customisations and added plugins, but the issue also occurs on the default LazyVim config. So, I could use some help on how to debug this further, or any ideas on what could be causing this.

EDIT: Turns out the minimal config does actually reproduce the issue, I just wasn't testing the correct case, so it's probably not related to the LazyVim config. Still could use any help in pinning down the cause of the issue; whether it's `nvim-cmp`, the Astro language server, or something else.

The "minimal" config I used for reference:
```Lua
local lazypath = vim.fn.stdpath("data") .. "/lazy/lazy.nvim"
if not vim.loop.fs_stat(lazypath) then
  -- bootstrap lazy.nvim
  -- stylua: ignore
  vim.fn.system({ "git", "clone", "--filter=blob:none", "https://github.com/folke/lazy.nvim.git", "--branch=stable", lazypath })
end
vim.opt.rtp:prepend(vim.env.LAZY or lazypath)

require("lazy").setup({
	spec = {
		{
			"williamboman/mason.nvim",
			cmd = "Mason",
			keys = { { "<leader>cm", "<cmd>Mason<cr>", desc = "Mason" } },
			build = ":MasonUpdate",
			opts = {
				ensure_installed = {
					"astro-language-server",
				},
			},
			---@param opts MasonSettings | {ensure_installed: string[]}
			config = function(_, opts)
				require("mason").setup(opts)
				local mr = require("mason-registry")
				local function ensure_installed()
					for _, tool in ipairs(opts.ensure_installed) do
						local p = mr.get_package(tool)
						if not p:is_installed() then
							p:install()
						end
					end
				end
				if mr.refresh then
					mr.refresh(ensure_installed)
				else
					ensure_installed()
				end
			end,
		},
		"hrsh7th/cmp-nvim-lsp",
		{
			"hrsh7th/nvim-cmp",
			version = false, -- last release is way too old
			event = "InsertEnter",
			dependencies = {
				"hrsh7th/cmp-nvim-lsp",
			},
			opts = function()
				vim.api.nvim_set_hl(0, "CmpGhostText", { link = "Comment", default = true })
				local cmp = require("cmp")
				local defaults = require("cmp.config.default")()
				return {
					completion = {
						completeopt = "menu,menuone,noinsert",
					},
					mapping = cmp.mapping.preset.insert({
						["<C-n>"] = cmp.mapping.select_next_item({ behavior = cmp.SelectBehavior.Insert }),
						["<C-p>"] = cmp.mapping.select_prev_item({ behavior = cmp.SelectBehavior.Insert }),
						["<C-b>"] = cmp.mapping.scroll_docs(-4),
						["<C-f>"] = cmp.mapping.scroll_docs(4),
						["<C-l>"] = cmp.mapping.complete(),
						["<C-e>"] = cmp.mapping.abort(),
						["<CR>"] = cmp.mapping.confirm({ select = true }), -- Accept currently selected item. Set `select` to `false` to only confirm explicitly selected items.
						["<S-CR>"] = cmp.mapping.confirm({
							behavior = cmp.ConfirmBehavior.Replace,
							select = true,
						}), -- Accept currently selected item. Set `select` to `false` to only confirm explicitly selected items.
					}),
					sources = cmp.config.sources({
						{ name = "nvim_lsp" },
					}),
					experimental = {
						ghost_text = {
							hl_group = "CmpGhostText",
						},
					},
					sorting = defaults.sorting,
				}
			end,
		},
		"williamboman/mason-lspconfig.nvim",
		{
			"neovim/nvim-lspconfig",
			event = { "BufReadPre", "BufNewFile" },
			dependencies = {
				"mason.nvim",
				"williamboman/mason-lspconfig.nvim",
				"hrsh7th/cmp-nvim-lsp",
			},
			---@class PluginLspOpts
			opts = {
				-- LSP Server Settings
				---@type lspconfig.options
				servers = {
					astro = {},
				},
				setup = {},
			},
			---@param opts PluginLspOpts
			config = function(_, opts)
				local servers = opts.servers
				local has_cmp, cmp_nvim_lsp = pcall(require, "cmp_nvim_lsp")
				local capabilities = vim.tbl_deep_extend(
					"force",
					{},
					vim.lsp.protocol.make_client_capabilities(),
					has_cmp and cmp_nvim_lsp.default_capabilities() or {}
				)

				local function setup(server)
					local server_opts = vim.tbl_deep_extend("force", {
						capabilities = vim.deepcopy(capabilities),
					}, servers[server] or {})

					if opts.setup[server] then
						if opts.setup[server](server, server_opts) then
							return
						end
					elseif opts.setup["*"] then
						if opts.setup["*"](server, server_opts) then
							return
						end
					end
					require("lspconfig")[server].setup(server_opts)
				end

				-- get all the servers that are available through mason-lspconfig
				local have_mason, mlsp = pcall(require, "mason-lspconfig")
				local all_mslp_servers = {}
				if have_mason then
					all_mslp_servers = vim.tbl_keys(require("mason-lspconfig.mappings.server").lspconfig_to_package)
				end

				local ensure_installed = {} ---@type string[]
				for server, server_opts in pairs(servers) do
					if server_opts then
						server_opts = server_opts == true and {} or server_opts
						-- run manual setup if mason=false or if this is a server that cannot be installed with mason-lspconfig
						if server_opts.mason == false or not vim.tbl_contains(all_mslp_servers, server) then
							setup(server)
						else
							ensure_installed[#ensure_installed + 1] = server
						end
					end
				end

				if have_mason then
					mlsp.setup({ ensure_installed = ensure_installed, handlers = { setup } })
				end
			end,
		},
	},
	defaults = {
		-- By default, only LazyVim plugins will be lazy-loaded. Your custom plugins will load during startup.
		-- If you know what you're doing, you can set this to `true` to have all your custom plugins lazy-loaded by default.
		lazy = false,
		-- It's recommended to leave version=false for now, since a lot the plugin that support versioning,
		-- have outdated releases, which may break your Neovim install.
		version = false, -- always use the latest git commit
		-- version = "*", -- try installing the latest stable version for plugins that support semver
	},
	performance = {
		rtp = {
			-- disable some rtp plugins
			disabled_plugins = {
				"gzip",
				-- "matchit",
				-- "matchparen",
				-- "netrwPlugin",
				"tarPlugin",
				"tohtml",
				"tutor",
				"zipPlugin",
			},
		},
	},
})
```

**Comment by abeldekat** [comment]
Hello @Trildar, 

Using stock `lazyvim starter`, I am trying to reproduce this behavior, on arch linux using neovim v 0.9.2

I added the following code to the starter template:

`lua/plugins/test.lua`:

``` lua
-- warning: could not load prettier or prettier-plugin-astro
return {
  {
    "nvim-treesitter/nvim-treesitter",
    opts = function(_, opts)
      vim.list_extend(opts.ensure_installed, {
        "astro",
      })
    end,
  },
  {
    "neovim/nvim-lspconfig",
    opts = {
      servers = {
        astro = {},
      },
    },
  },
}
```

As I am not familiar with `astro`, I just used the instructions for starting a project:
`npm create astro@latest`

I answered 'yes' to all questions the interactive installer asked.

Now, I have a project, containing file `src/layouts/Layout.astro`
I will use the following code fragment in that file:
```
	<body>
		<slot />
	</body>
``` 

### Test 1
Below the line containing `slot`, start typing:
`<div id={t`
The completion engine offers `title`, a `variable`. I am able to accept the suggestion and type the line intended:
`<div id={title}></div>`

### Test 2
Add a space between `id` and `{`, start typing:
`<div id ={t`
The result is not correct:
`<div id =title`

### Test 3
Using the result of the first test, adding a space between `id` and `=`:
`<div id ={title}></div>`
Inside the placeholder, `ciw`. Now type `t`. The completion for title appears, and accept.
The result is not correct:
`<div id =title}></div>`

### Test 4
All previous tests, but instead of accepting the `title` variable from the `lsp`, accept `title` text from buffer completion. Now all results are correct.

### Evaluating

When repeating the tests with the minimal config you provided, the results are the same.

It's quite possible I missed something obvious. Based on the tests above, I think the problem occurs when adding a space, as shown in test 2. Perhaps, the behavior is caused by either the `lsp` or `cmp-nvim-lsp`.

If the test cases above do not cover your use case, a sample `.astro` file would be helpful.

Best regards! 

edit: Added test 4, changed evaluation.

**Reply by Trildar** [reply]
Seems I didn't do my testing quite right. I was able to reproduce the cases you mention, both on the minimal config and also on my current LazyVim one, but the behaviour I was experiencing is a little different. At least one case it occurs is when the tag is multiline, e.g.

```
<input
    id={}
/>
```

Also, this issue actually does occur on the minimal config as well. I was just mistaken cause I was actually testing a subtly different case where there was no issue (basically what you got in test 1).

**Reply by dpetka2001** [reply]
I was able to reproduce it with a fresh Lazyvim starter template. It only happens in multiline tags for me. When the property was on the same line as the name of the tag, it wasn't happening. I also tried disabling the Astro LSP server and only accepting suggestions from text buffer and there wasn't any problem with that even on multiline tags. Then I tried with Astro LSP server enabled, but instead of `nvim-cmp` autocompletion I tried with `<C-x><C-o>` Neovim omnifunc completion and it was behaving normally. So, my guess would be that it has something to do between `cmp-nvim-lsp` responsible for the LSP sources and Astro LSP itself, because to be honest I didn't notice any such behavior with other LSPs (at least in my experience except for the `codeium.nvim` suggestions in my case, but that was also a problem with the source itself I believe instead of `nvim-cmp`).

**Reply by dpetka2001** [reply]
On second thought, this must have something to do with `nvim-cmp` or `cmp-nvim-lsp` whichever is responsible for calculating the completion range. Since the default Neovim omnifunc which is set to `server.vim.lsp.omnifunc` according to the LSP attached, also must be communicating directly with the LSP and doesn't have this problem. It must be something about how `nvim-cmp` handles `props` accross multiline tags? Does it also happen for other kind of autocompletion that aren't `props`?  Obviously, just guessing based on observations.

**Reply by abeldekat** [reply]
As `omnifunc` completion is behaving as expected, perhaps creating an issue in `nvim-cmp` is appropriate.

It seems that the accepted word is inserted one column too far:

```
<input
    id={title}
/>
```

Cursor in title, `ciw`, accept the `title` variable:
```
<input
    id={ttitle
/>
```

There is an issue regarding astro already in the repo of nvim-cmp: 1471, Completion doesn't close in .astro files

**Reply by Trildar** [reply]
Yeah, I'll open up an issue on `nvim-cmp`. Some additional notes:

You're right that it seems to start one character too late, at least usually, I'm pretty sure I have seen it leave 2 characters out of what you type as well, but it's rarer and I'm not sure how to trigger it. If you put more text after on the same line though, you'll likely find that it overwrites more than one character past where it should.

Similar behaviour occurs in nested interpolations even when on one line, e.g.

```
{<div id={test}></div>}
```

Not too sure if the existing issue is related. The original post is about getting an unexpected completion prompt immediately after confirming a completion. The first completion also looks like it comes from some snippet rather than Astro LSP.

**Reply by abeldekat** [reply]
The issue in `nvim-cmp`: [Astro LSP completions sometimes write over incorrect ranges ](https://github.com/hrsh7th/nvim-cmp/issues/1705)

**Comment by dpetka2001** [comment]
I noticed this also at random moments a while ago, but didn't pay further attention as it wasn't happening consistently. It happened to me in Lua files, so I guess it's not completely Astro related. The last week it didn't happen to me. The only thing I changed in my config, is I removed `codeium.nvim` since it was archived. I added `codeium.vim` instead, but added 2 extra events on `nvim-cmp` to not show the virtual text when the autocompletion menu is open. Are you by any chance using `copilot` or `codeium`? It might have to do something with those sources and `nvim-cmp`. Could you check if disabling either of them fixes your issue?

**Reply by Trildar** [reply]
Might or might not be related by something in `nvim-cmp`. I'm not using `copilot` or `codeium` though. My customised config doesn't have those, and I can reproduce this on the default LazyVim config, and the cut down config I put in the original post that basically just has `nvim-cmp`, `mason`, and supporting plugins to get LSP completions.
